decisions/event-listeners-and-state.md

```diff
@@ -0,0 +1,83 @@
+# Event listeners and state
+
+## TL;DR
+
+Adding default listeners and changing visual state inside our components should be done with caution and avoided when possible, as it can cause unexpected problems especially if consumers extend our components.
+
+## Decision
+
+When creating a component that changes its state based on user interaction (e.g. press/click) we should avoid adding listeners and changing the state inside the component whenever possible.
+
+Preference should be given to exposing the state change logic via a function and letting users add the listeners as they see fit.
+
+Consider the following example:
+
+```Kotlin
+// Bad
+class BpkChip {
+  init {
+    setOnClickListener({
+      if (!disabled) {
+        selected = !selected
+      }
+    })
+  }
+}
+
+// Good
+class BpkChip {
+  fun toggle() {
+    if (!disabled) {
+      selected = !selected
+    }
+  }
+}
+
+BpkChip().setClickListener { it.toggle() }
+```
+
+## Thinking
+
+There are two main problems with adding the listener by default.
+
+**Extension and order of initialization:**
+
+Consider the example above again but now with a class extending our component and overriding the click listener. *Comments show the order of initialization.*
+
+```Kotlin
+class BpkChip {
+  init {
+    // 1
+    setOnClickListener({
+      if (!disabled) {
+        selected = !selected
+      }
+    })
+  }
+}
+
+class MyCustomChip {
+  // 3
+  val myAnalyticProp = AnalyticProp()
+ 
+  override fun setOnClickListener(l: Listener) {
+    // 2
+    super.setOnClickListener(WrapWithAnalitics(l, myAnalyticProp))
+  }
+}
+
+```
+
+In the example above `setOnClickListener` will be called by the super class before the props and init block have been called in the child class. This can be very hard to identify, especially from a user perspective.
```

Comment:
```suggestion
In the example above `setOnClickListener` will be called by the super class before the props and `init block` have been called in the child class. This can be very hard to identify, especially from a user perspective.
```